### PR TITLE
implement pytorch shim with benchmark tests

### PR DESCRIPTION
**Pull request by gnieuwenhuis**
In order to enable deeper experiments, creating a simple shim that defined a pytorch model instead of an sklearn MLP architecture.

The goal here is to create a nearly identical model using pytorch, including unit tests to ensure similarity between the two models.

### CODE REVIEW

File: mermaid_classifier/pyspacer/torch_classifier.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        # from `self.random_state` every time, so identical input + same
+        # `random_state` reproduces the same shuffle across partial_fit
+        # calls. Callers that want per-call variation should vary
+        # random_state themselves (as MermaidTrainer does implicitly by
+        # passing different batches).
+        base_seed = self.random_state
+        if base_seed is None:
+            return np.random.default_rng()
+        return np.random.default_rng(int(base_seed))
+
```
[the remark this answers is not in the document]
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/torch_classifier.py:148]: Fixed in `_seed_rng()`. For `random_state=None` the shuffle RNG is now seeded once per instance from NumPy's global RNG and reused across `partial_fit` calls, so `np.random.seed(...)` makes shuffling reproducible (matching `check_random_state(None)`'s use of the global singleton) while still advancing between calls. Int `random_state` keeps the per-call re-seed for repeatable-per-call shuffles. Note weight init for the `None` case stays torch-random — sklearn parity here is about the shuffle RNG, which is what now honours the global seed.

File: mermaid_classifier/pyspacer/torch_classifier.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        if not hasattr(self, "_module"):
+            raise RuntimeError(
+                "TorchMLPClassifier is not fitted. Call partial_fit or fit"
+                " before predict/predict_proba."
+            )
+        X_arr = np.asarray(X, dtype=np.float32)
+        self._module.eval()
+        with torch.no_grad():
+            probs = F.softmax(self._module(torch.from_numpy(X_arr)), dim=1)
+        return probs.numpy().astype(np.float64)
```
[the remark this answers is not in the document]
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/torch_classifier.py:305]: Fixed. `_forward_probs()` now validates `X` is 2D with the fitted `n_features_in_` count before running the module, raising an actionable `ValueError` instead of an opaque softmax crash (mirrors the checks already in `partial_fit`).

File: tests/pyspacer/test_mlp_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+
+
+SEED = 42
+N_CLASSES = 5
+N_FEATURES = 32
+N_TRAIN = 500
+N_VAL = 200
+HIDDEN = (64,)
+LR = 1e-2
+EPOCHS = 20
```
[the remark this answers is not in the document]
Comment (gnieuwenhuis) [reply tests/pyspacer/test_mlp_benchmark.py:37]: Leaving as-is. Measured the full suite at ~2.2–2.6s for all 30 tests locally (synthetic data is small), so it isn't a meaningful CI cost. Moving training into `setUpClass` would share one mutable model across tests, but several of them further-train, pickle, or run `partial_fit` with different chunk sizes (`test_partial_fit_accumulates_across_calls`), so caching would trade away test isolation for negligible wall-clock gain.

File: scripts/classifier_train.py [hunk 4]
```diff
@@ -22,7 +22,7 @@
         dataset_options=DatasetOptions(
             # Specifying False here means you're only training on CoralNet sources.
             include_mermaid=False,
-            coralnet_sources_csv='../sagemaker/sources/OldCoralNetSourcesToKeep.csv',
+            coralnet_sources_csv='../sagemaker/sources/CoralNetSourcesFirst10.csv',
```
[the remark this answers is not in the document]
Comment (gnieuwenhuis) [reply scripts/classifier_train.py]: Reverted to the previous default (`OldCoralNetSourcesToKeep.csv`). It was an unrelated local-experiment tweak in this scratch script and shouldn't ride along with the shim PR.

File: tests/pyspacer/test_mlp_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+
+    # --- Tests -----------------------------------------------------------
+
+    def test_converges_on_training_set(self):
+        """Training accuracy should exceed 85% on the easy synthetic task."""
+        clf = self._make_classifier()
+        self._train(clf)
+        preds = clf.predict(self.X_train)
+        acc = float(np.mean(preds == self.y_train))
+        self.assertGreater(
```
Comment (yeelauren) [inline tests/pyspacer/test_mlp_benchmark.py:140]: Where is 85% coming from ? Have you considered adding tests that compare the outputs directly from scikit and pytorch instead of a greater than ?

File: tests/pyspacer/test_mlp_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+        self._train(clf)
+        preds = clf.predict(self.X_val)
+        acc = float(np.mean(preds == self.y_val))
+        self.assertGreater(
+            acc, 0.80,
+            f"Validation accuracy {acc:.3f} below threshold 0.80"
+            f" for {type(clf).__name__}"
+        )
+
+    def test_predict_proba_shape_and_normalisation(self):
```
Comment (yeelauren) [inline tests/pyspacer/test_mlp_benchmark.py:158]: What about the actual probA values ?

File: tests/pyspacer/test_mlp_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+        probs_after = clf2.predict_proba(self.X_val)
+
+        np.testing.assert_array_equal(preds_before, preds_after)
+        np.testing.assert_allclose(probs_before, probs_after,
+                                   rtol=1e-5, atol=1e-6)
+
+    def test_partial_fit_accumulates_across_calls(self):
+        """Multiple small partial_fit calls should still converge."""
+        clf = self._make_classifier()
+        rng = np.random.RandomState(SEED + 2)
```
Comment (yeelauren) [inline tests/pyspacer/test_mlp_benchmark.py]: why SEED +2 ?

File: tests/pyspacer/test_mlp_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+        # Deliberately tiny chunks: tests the incremental-fit contract.
+        train_via_partial_fit(
+            clf, self.X_train, self.y_train, self.classes_list,
+            epochs=EPOCHS, chunk_size=50, rng=rng,
+        )
+        acc = float(np.mean(clf.predict(self.X_val) == self.y_val))
+        self.assertGreater(
+            acc, 0.75,
```
Comment (yeelauren) [inline tests/pyspacer/test_mlp_benchmark.py:224]: same why were these values chosen ?

File: tests/pyspacer/test_mlp_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+        X_train, y_train, X_val, y_val = train_val_split(
+            X_all, y_all, n_val=N_VAL)
+        classes_list = sorted(np.unique(y_all).tolist())
+
+        clf = TorchMLPClassifier(
+            hidden_layer_sizes=HIDDEN,
+            learning_rate_init=LR,
+            random_state=SEED,
+        )
+        rng = np.random.RandomState(SEED + 1)
```
Comment (yeelauren) [inline tests/pyspacer/test_mlp_benchmark.py]: SEED + 1

File: pyproject.toml [hunk 10]
```diff
@@ -38,6 +38,9 @@ pyspacer = [
     "s3fs",
     # pyspacer uses scikit-learn, but we also use it directly for metrics.
     "scikit-learn",
+    # torch is already a transitive dep via pyspacer, but we import it
+    # directly in torch_classifier.py so declare it here too.
+    "torch",
```
Comment (yeelauren) [inline pyproject.toml:43]: maybe out of scope for this PR, but do any of these need to be pinned ?

**Review by yeelauren — CHANGES_REQUESTED** [review]
Thanks Greg, this moves in a great direction for pyspacer, and PyTorch has many, many benefits. 

I left a few comments re: values that were chosen and I think some tests could include the calibrated probabilies between both - unless i missed it. 

One other consideration is that i believe RNG behaves differently between pytorch, numpy and sklearn and to double check how data may be loaded or RNG may be invoked.

- https://docs.pytorch.org/docs/2.12/notes/randomness.html

- https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/ 

- https://learnopencv.com/ensuring-training-reproducibility-in-pytorch/

File: pyproject.toml (hunk 10 above)
Comment (gnieuwenhuis) [reply pyproject.toml:43]: The current pattern is that pyspacer pins torch, and we pin pyspacer, which is intentional to avoid version mismatches, especially when trying to keep backwards compatibility.

File: tests/pyspacer/test_mlp_benchmark.py (hunk 7 above)
Comment (gnieuwenhuis) [reply tests/pyspacer/test_mlp_benchmark.py]: It is on purpose, but I understand the lack of readability - the main idea is that SEED generates teh dataset, SEED + 1 drives the training shuffle, SEED + 2 drives the tiny-chunk shuffle. All of this is to avoid things like the correlation of shuffle order with data generation. I will add a short comment to make it clear this is on purpose.

File: tests/pyspacer/test_mlp_benchmark.py (hunk 9 above)
Comment (gnieuwenhuis) [reply tests/pyspacer/test_mlp_benchmark.py]: see my comment above.

File: tests/pyspacer/test_mlp_benchmark.py (hunk 5 above)
Comment (gnieuwenhuis) [reply tests/pyspacer/test_mlp_benchmark.py:140]: These values (including the 75% below) are chosen as loose sanity floors that answer the question "is the model capable of learning?". The `make_gaussian_clusters` function builds well seperated clusters (centroids at radius ~3, std 1.3) that the model should comfortably perform well on, so the sanity floors are set well below the expected performance (~95%) to avoid flaky tests while still ensuring the model is capable of learning something.

The head to head comparisons live in `MLPParityTest`, while these are deliberately loose convergence floors. I'll add a short comment to that effect as well for readability.

File: tests/pyspacer/test_mlp_benchmark.py (hunk 6 above)
Comment (gnieuwenhuis) [reply tests/pyspacer/test_mlp_benchmark.py:158]: There is a test `test_torch_predict_proba_distribution_close` that checks the argmax is in >85% agreement, but you're right there is not a specific test that checks probA values, and also doesn't check post-calibration probablilities. I've added tests accordingly.

File: tests/pyspacer/test_mlp_benchmark.py (hunk 8 above)
Comment (gnieuwenhuis) [reply tests/pyspacer/test_mlp_benchmark.py:224]: see above comment, I'll add a comment in the code fro readability

**Review by yeelauren — APPROVED** [review]
